File: scripts/oop-practice-sources/csharp/animal-inheritance.cs
class Animal
{
    public string Name { get; }

    public Animal(string name)
    {
        Name = name;
    }

    public void Eat()
    {
        Console.WriteLine($"{Name} ест");
    }
}

class Cat : Animal
{
    public Cat(string name) : base(name) { }

    public void Speak()
    {
        Console.WriteLine("Мяу!");
    }
}

class Dog : Animal
{
    public Dog(string name) : base(name) { }

    public void Speak()
    {
        Console.WriteLine("Гав!");
    }
}

class Program
{
    static void Main()
    {
        var cat = new Cat("Мурка");
        var dog = new Dog("Шарик");
        cat.Eat();
        cat.Speak();
        dog.Eat();
        dog.Speak();
    }
}

File: scripts/oop-practice-sources/rust/animal-inheritance.rs
struct Animal {
    name: String,
}

impl Animal {
    fn eat(&self) {
        println!("{} ест", self.name);
    }
}

struct Cat {
    animal: Animal,
}

impl Cat {
    fn new(name: &str) -> Self {
        Self {
            animal: Animal {
                name: name.to_string(),
            },
        }
    }

    fn eat(&self) {
        self.animal.eat();
    }

    fn speak(&self) {
        println!("Мяу!");
    }
}

struct Dog {
    animal: Animal,
}

impl Dog {
    fn new(name: &str) -> Self {
        Self {
            animal: Animal {
                name: name.to_string(),
            },
        }
    }

    fn eat(&self) {
        self.animal.eat();
    }

    fn speak(&self) {
        println!("Гав!");
    }
}

fn main() {
    let cat = Cat::new("Мурка");
    let dog = Dog::new("Шарик");
    cat.eat();
    cat.speak();
    dog.eat();
    dog.speak();
}
